### ISSUE TITLE
Performance differences caused by image size

### ISSUE THREAD

**Issue by ohgyeongsu** [ultralytics/ultralytics#20859]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I trained the model with an image size `imgsz` of 1440.
However, when I run predictions, the model performs better with an image size of 1280 compared to 1440.
Could you explain why reducing the image size during inference leads to better performance?

Could my preprocessing be the problem?
Here is the relevant code.

The original image size is w=2448, h=2048.
I want to crop the image using [388, 986, 1690, 969] in xywh format.
`rect` was set to False during both training and prediction.

```
import math
from pathlib import Path

import cv2
import numpy as np
from ultralytics import YOLO, settings
from ultralytics.data.dataset import Instances, YOLODataset
from ultralytics.models.yolo.detect import DetectionTrainer
from ultralytics.utils import LOGGER, colorstr
from ultralytics.utils.ops import resample_segments
from ultralytics.utils.torch_utils import de_parallel


class CustomTrainer(DetectionTrainer):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def build_dataset(self, img_path, mode="train", batch=None):
        gs = max(int(de_parallel(self.model).stride.max() if self.model else 0), 32)

        # Create and return custom dataset
        return CustomCropDataset(
            img_path=img_path,
            imgsz=self.args.imgsz,
            batch_size=batch,
            augment=mode == "train",
            hyp=self.args,
            rect=self.args.rect or mode == "val",
            cache=self.args.cache or None,
            single_cls=self.args.single_cls or False,
            stride=int(gs),
            pad=0.0 if mode == "train" else 0.5,
            prefix=colorstr(f"{mode}: "),
            task=self.args.task,
            classes=self.args.classes,
            data=self.data,
            fraction=self.args.fraction if mode == "train" else 1.0,
        )


class CustomCropDataset(YOLODataset):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.ori_h, self.ori_w = (2048, 2448)
        self.coordinate = [388, 986, 1690, 969]

    def load_image(self, i, rect_mode=True):
        im, f, fn = self.ims[i], self.im_files[i], self.npy_files[i]
        if im is None:  # not cached in RAM
            if fn.exists():  # load npy
                try:
                    im = np.load(fn)
                except Exception as e:
                    LOGGER.warning(f"{self.prefix}WARNING ⚠️ Removing corrupt *.npy image file {fn} due to: {e}")
                    Path(fn).unlink(missing_ok=True)
                    im = cv2.imread(f)  # BGR
            else:  # read image
                im = cv2.imread(f)  # BGR
            if im is None:
                raise FileNotFoundError(f"Image Not Found {f}")

            roi_x, roi_y, roi_w, roi_h = self.coordinate
            crop_im = im[roi_y : roi_y + roi_h, roi_x : roi_x + roi_w]

            h0, w0 = crop_im.shape[:2]
            if rect_mode:
                r = self.imgsz / max(h0, w0)  # ratio
                if r != 1:  # if sizes are not equal
                    w, h = (min(math.ceil(w0 * r), self.imgsz), min(math.ceil(h0 * r), self.imgsz))
                    crop_im = cv2.resize(crop_im, (w, h), interpolation=cv2.INTER_LINEAR)
            elif not (h0 == w0 == self.imgsz):
                crop_im = cv2.resize(crop_im, (self.imgsz, self.imgsz), interpolation=cv2.INTER_LINEAR)

            # Add to buffer if training with augmentations
            if self.augment:
                self.ims[i], self.im_hw0[i], self.im_hw[i] = (
                    crop_im,
                    (h0, w0),
                    crop_im.shape[:2],
                )  # im, hw_original, hw_resized
                self.buffer.append(i)
                if 1 < len(self.buffer) >= self.max_buffer_length:  # prevent empty buffer
                    j = self.buffer.pop(0)
                    if self.cache != "ram":
                        self.ims[j], self.im_hw0[j], self.im_hw[j] = None, None, None

            return crop_im, (h0, w0), crop_im.shape[:2]

        return self.ims[i], self.im_hw0[i], self.im_hw[i]

    def update_labels_info(self, label):
        bboxes = label.pop("bboxes")
        segments = label.pop("segments", [])
        keypoints = label.pop("keypoints", None)
        bbox_format = label.pop("bbox_format")
        normalized = label.pop("normalized")

        x_min, y_min, w_crop, h_crop = self.coordinate

        if normalized:
            bboxes[:, 0] *= self.ori_w
            bboxes[:, 1] *= self.ori_h
            bboxes[:, 2] *= self.ori_w
            bboxes[:, 3] *= self.ori_h

        bboxes[:, 0] -= x_min
        bboxes[:, 1] -= y_min

        bboxes[:, 0] *= label["ratio_pad"][1]
        bboxes[:, 1] *= label["ratio_pad"][0]
        bboxes[:, 2] *= label["ratio_pad"][1]
        bboxes[:, 3] *= label["ratio_pad"][0]

        if normalized:
            bboxes[:, 0] /= label["resized_shape"][1]
            bboxes[:, 1] /= label["resized_shape"][0]
            bboxes[:, 2] /= label["resized_shape"][1]
            bboxes[:, 3] /= label["resized_shape"][0]

        if len(segments) > 0:
            if normalized:
                segments = [s * np.array([self.ori_w, self.ori_h]) for s in segments]
            segments = [s - np.array([x_min, y_min]) for s in segments]
            if normalized:
                segments = [s / np.array([w_crop, h_crop]) for s in segments]

        if keypoints is not None:
            if normalized:
                keypoints[:, 0] *= self.ori_w
                keypoints[:, 1] *= self.ori_h
            keypoints[:, 0] -= x_min
            keypoints[:, 1] -= y_min
            if normalized:
                keypoints[:, 0] /= w_crop
                keypoints[:, 1] /= h_crop

        segment_resamples = 100 if self.use_obb else 1000
        if len(segments) > 0:
            max_len = max(len(s) for s in segments)
            segment_resamples = (max_len + 1) if segment_resamples < max_len else segment_resamples
            segments = np.stack(resample_segments(segments, n=segment_resamples), axis=0)
        else:
            segments = np.zeros((0, segment_resamples, 2), dtype=np.float32)

        label["instances"] = Instances(bboxes, segments, keypoints, bbox_format=bbox_format, normalized=normalized)
        return label

    def __getitem__(self, index):
        return self.transforms(self.get_image_and_label(index))


class YOLOv8:
    def __init__(self, config: dict):
        self.config = config
        settings.update({"mlflow": True})
        self.model = YOLO(**self.config)

    def train(self, config: dict):
        self.model.train(trainer=CustomTrainer, **config)

    def tune(self, space, config: dict):
        return self.model.tune(space=space, **config)

    def predict(self, source, device, config: dict):
        # visualize
        return self.model.predict(source=source, device=device, **config)

    def val(self):
        return self.model.val()

    def export(self, format, device, opset, config: dict):
        return self.model.export(format=format, device=device, opset=opset, **config)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ohgyeongsu, thank you for reaching out and for providing detailed information about your workflow and code! 🚀 This is an automated response to help you get started. An Ultralytics engineer will also review your issue and assist you further soon.

For new users, we recommend checking out the [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides, where many frequently asked questions are already answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) if possible. This will help us investigate the behavior you are observing with different `imgsz` values during inference.

If this is a custom training ❓ Question, please include relevant dataset image examples and training logs, and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Feel free to join the Ultralytics community:
- For real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- To share knowledge and experiences, check out our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Please ensure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

We appreciate your detailed code snippet and description! Please provide a minimum reproducible example if you are seeing unexpected behavior, and an engineer will follow up with tailored assistance soon.

**Comment by glenn-jocher** [comment]
The performance difference you're experiencing is likely due to how your custom cropping implementation handles image resizing at different target sizes. Looking at your `load_image` method, there's a potential issue with the resizing logic.

When you train at 1440 but inference performs better at 1280, this suggests your cropped region (1302x969 pixels from your coordinate [388, 986, 1690, 969]) is being resized differently than expected. At 1440, the aspect ratio preservation might be creating a different effective resolution than at 1280.

Try using the standard YOLO11 [LetterBox transform](https://docs.ultralytics.com/reference/data/augment/#ultralytics.data.augment.LetterBox) instead of your custom resizing logic to ensure consistent preprocessing between training and inference. You can integrate cropping with LetterBox by cropping first, then applying the standard transform pipeline.

Also consider that your current `rect_mode=True` logic differs from the standard BaseDataset implementation - you might want to align it more closely with the [BaseDataset.load_image method](https://docs.ultralytics.com/reference/data/base/#load_image) for consistency.

**Comment by WenLinLliu** [comment]
I also meet this question！But Idont why

**Comment by glenn-jocher** [comment]
@WenLinLliu This typically happens when there's inconsistent preprocessing between training and inference - as discussed above, the issue is often related to how custom resizing/cropping logic interacts with YOLO's standard LetterBox transforms, especially around ratio_pad corrections being applied in the wrong order or twice.
